src/coverage/coverageService.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Espressif Systems (Shanghai) CO LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { basename, join } from "path";
+import * as vscode from "vscode";
+import { appendIdfAndToolsToPath, execChildProcess } from "../utils";
+import { OutputChannel } from "../logger/outputChannel";
+import { Logger } from "../logger/logger";
+
+export interface countRange {
+  range: vscode.Range;
+  count: Number;
+}
+
+export interface textEditorWithCoverage {
+  coveredLines: vscode.Range[];
+  partialLines: vscode.Range[];
+  uncoveredLines: vscode.Range[];
+  editor: vscode.TextEditor;
+  countPerLines: countRange[];
+  allLines: vscode.Range[];
+}
+
+export async function buildJson(dirPath: string) {
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      "/Users/brian/esp-idf/components/",
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--json",
+    ],
+    dirPath
+  );
+  return JSON.parse(result);
+}
+
+export async function buildHtml(dirPath: string) {
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      "/Users/brian/esp-idf/components/",
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--html",
+    ],
+    dirPath
+  );
+  return result;
+}
+
+function _runCmd(cmd: string, args: string[], dirPath: string) {
+  const modifiedEnv = appendIdfAndToolsToPath();
+  return execChildProcess(
+    `${cmd} ${args.join(" ")}`,
+    dirPath,
+    OutputChannel.init(),
+    { env: modifiedEnv, cwd: dirPath }
+  ).catch((e) => {
+    const msg = e.message ? e.message : e;
+    Logger.error("Error on gcov cmd.\n" + msg, e);
+    OutputChannel.appendLine("Error building gcov cmd.\n" + msg);
+    return "";
```

Comment:
maybe can use `spawn`? 
Although I am not sure since how long the command responses are ? if they are huge, `exec ` will wait for all of the response and return a buffer instead of stream!!

Reply:
I had tried with just a couple of files so exec worked pretty well but yeah if the user chooses many sources files for coverage it could truncate the output. Will fix now.